engine/execution/ingestion/uploader/uploader.go

```diff
@@ -26,74 +27,99 @@ func NewAsyncUploader(uploader Uploader,
 	maxRetryNumber uint64,
 	log zerolog.Logger,
 	metrics module.ExecutionMetrics) *AsyncUploader {
-	return &AsyncUploader{
-		unit:                engine.NewUnit(),
+	a := &AsyncUploader{
 		uploader:            uploader,
 		log:                 log.With().Str("component", "block_data_uploader").Logger(),
 		metrics:             metrics,
 		retryInitialTimeout: retryInitialTimeout,
 		maxRetryNumber:      maxRetryNumber,
+		queue:               make(chan *execution.ComputationResult, 100),
```

Comment:
```suggestion
		// we use a channel rather than a Fifoqueue here because a Fifoqueue might drop items when full,
		// but it is not acceptable to skip uploading an execution result
		queue:               make(chan *execution.ComputationResult, 100),
```
Let's document why we're choosing a channel here.

Reply:
Regarding the channel size:
- I believe that block execution will [block until an `Upload` call completes](https://github.com/onflow/flow-go/blob/231d260e94216d835501b5608cf329cef9b5f40f/engine/execution/ingestion/core.go#L324-L399) (uploading is invoked concurrently, but the outer function waits for completion before returning). I didn't look into it enough to be 100% sure though - hopefully Leo can speak to this.
- The current behaviour is that each `Upload` call will spawn a goroutine (within `AsyncUploader`), so will never block execution (instead it will keep increasing # of goroutines until we run out of memory)
- This PR changes the behaviour so that, when the channel becomes full, execution will get blocked
- Stopping execution is bad, so we should set a relatively large buffer for pending execution results. 5000 is about an hour's worth of blocks, which seems reasonable to me.

Reply:
Is this refactor to deprecate Engine.Unit? 

If so, we better have a equivalent replacement for the `Unit.Launch` method, which spawn a goroutine to make a process non-blocking (async). We are replacing with a queue with a fixed buffer size, which is similar but not equivalent, as Jordan mentioned, when the buffer size is full, it will block the block execution. 

Note, we are actually deprecating the AsyncUploader, both mainnet and testnet have stopped uploading block data to GCP, so I'm OK that we replace it with a fixed size queue. But my point is more for finding an equivalent replacce for Unit.Launch which is used in many other places.

Reply:
Yes, the referenced issue is about deprecating Engine.Unit. From what I understood one of the reasons to move away from Unit.Launch is to get more control over, and insight into, resource usage.
The engine.Unit controls the amount of concurrent goroutines via a waitgroup, and any additional goroutines will wait until one completes before it begins working; ComponentManager instead declares worker goroutines upfront, and often(?) uses a fifoqueue to distribute work among them to be run asynchronously. In both cases we limit the amount of concurrency, but with a ComponentManager, resources for the waiting tasks are used in the queue where it's easier to measure them, rather than in a goroutine stack; it also lets us set a resource limit beyond which we can raise an error, if resource usage becomes unreasonable.
For example in my first PR, #6747, Unit.Launch() was replaced by a fifoqueue (with metrics) and worker started by a ComponentManager.

Jordan could probably explain it better if I got any of that wrong.

Reply:
Thanks for the comment - I agree with Tim's explanation of the motivation behind removing `engine.Unit`. Although, the waitgroup in `engine.Unit` is only for tracking ongoing goroutines (so we know when we are `Done()`); new goroutines will start immediately regardless of how many others are already in-progress.

> If so, we better have a equivalent replacement for the Unit.Launch method, which spawn a goroutine to make a process non-blocking (async)

@zhangchiqing The `ComponentManager`, and our transition from `engine.Unit` to the `ComponentManager`, is in part to avoid the pattern of using `Unit.Launch`. From my perspective, there is not a behaviourally equivalent replacement because we want to avoid unbounded creation of goroutines going forward as a pattern. The desired replacement is using a fixed number of component-scoped goroutines, which pull work from fixed-size queue, like Tim has done here.